### ISSUE TITLE
Datasets with different aspect ratios vs shuffling of dataset

### ISSUE THREAD

**Issue by shayanalibhatti** [ultralytics/yolov5#932]
Hi,

Great work developing yolov5. I have a question. Imagine you are combining different datasets, such as COCO (which has images of different aspect ratios), then some other datasets with image of same aspect ratio among that dataset. 

Wouldn't rectangular training (which sorts images by aspect ratio) hurt the shuffling of dataset ? as it would sort the images of same aspect ratio to be together in one batch. Thus, it might sort COCO first and then the other datasets in a serial manner. Or would just turning rectangular training OFF and shuffling dataset, do the job so that our model generalizes even on different datasets with images of different aspect ratios ?

Can you elaborate on this and correct my understanding?

**Comment by glenn-jocher** [comment]
With default settings this is a non-issue.

**Comment by glenn-jocher** [comment]
@shayanalibhatti good news 😃! Your original issue may now be fixed ✅ in PR #5623 by @werner-duvaud. This PR turns on shuffling in the YOLOv5 training DataLoader by default, which was missing until now. This works for all training formats: CPU, Single-GPU, Multi-GPU DDP.

```python
train_loader, dataset = create_dataloader(train_path, imgsz, batch_size // WORLD_SIZE, gs, single_cls,
                                          hyp=hyp, augment=True, cache=opt.cache, rect=opt.rect, rank=LOCAL_RANK,
                                          workers=workers, image_weights=opt.image_weights, quad=opt.quad,
                                          prefix=colorstr('train: '), shuffle=True)  # <--- NEW
```

I evaluated this PR against master on VOC finetuning for 50 epochs, and the results show a slight improvement in most metrics and losses, particularly in objectness loss and mAP@0.5, perhaps indicating that the shuffle addition may help delay overtraining.

https://wandb.ai/glenn-jocher/VOC
<img width="979" alt="Screenshot 2021-11-13 at 13 03 26" src="https://user-images.githubusercontent.com/26833433/141643075-70b06810-4727-4f5c-b03a-f68c34fe6934.png">

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by shayanalibhatti** [comment]
@glenn-jocher Thanks. Glad to know my insight was of help and will be, to the community. I am not working on computer vision related project right now so I cant test it but your observation of improvement in results is great. Keep up the great work.

**Comment by yizweithree** [comment]
> @shayanalibhatti good news 😃! Your original issue may now be fixed ✅ in PR #5623 by @werner-duvaud. This PR turns on shuffling in the YOLOv5 training DataLoader by default, which was missing until now. This works for all training formats: CPU, Single-GPU, Multi-GPU DDP.
> 
> ```python
> train_loader, dataset = create_dataloader(train_path, imgsz, batch_size // WORLD_SIZE, gs, single_cls,
>                                           hyp=hyp, augment=True, cache=opt.cache, rect=opt.rect, rank=LOCAL_RANK,
>                                           workers=workers, image_weights=opt.image_weights, quad=opt.quad,
>                                           prefix=colorstr('train: '), shuffle=True)  # <--- NEW
> ```
> 
> I evaluated this PR against master on VOC finetuning for 50 epochs, and the results show a slight improvement in most metrics and losses, particularly in objectness loss and mAP@0.5, perhaps indicating that the shuffle addition may help delay overtraining.
> 
> https://wandb.ai/glenn-jocher/VOC <img alt="Screenshot 2021-11-13 at 13 03 26" width="979" src="https://user-images.githubusercontent.com/26833433/141643075-70b06810-4727-4f5c-b03a-f68c34fe6934.png">
> 
> To receive this update:
> 
> * **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
> * **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
> * **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  [![Open In Colab](https://camo.githubusercontent.com/84f0493939e0c4de4e6dbe113251b4bfb5353e57134ffd9fcab6b8714514d4d1/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Kaggle](https://camo.githubusercontent.com/a08ca511178e691ace596a95d334f73cf4ce06e83a5c4a5169b8bb68cac27bef/68747470733a2f2f6b6167676c652e636f6d2f7374617469632f696d616765732f6f70656e2d696e2d6b6167676c652e737667)](https://www.kaggle.com/models/ultralytics/yolov5)
> * **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image [![Docker Pulls](https://camo.githubusercontent.com/280faedaf431e4c0c24fdb30ec00a66d627404e5c4c498210d3f014dd58c2c7e/68747470733a2f2f696d672e736869656c64732e696f2f646f636b65722f70756c6c732f756c7472616c79746963732f796f6c6f76353f6c6f676f3d646f636b6572)](https://hub.docker.com/r/ultralytics/yolov5)
> 
> Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

How to use shuffel in this new Update? or is shuffel turned on by default ?

**Comment by glenn-jocher** [comment]
@yizweithree shuffle is enabled by default now for training sets.

**Comment by yangsiyu007** [comment]
@glenn-jocher thanks for updating this and other threads on the issue! Can you elaborate more on the rationale for sorting the images by aspect ratio when doing "rectangular" training? With the above PR, images are still sorted by aspect ratio during evaluation on the val set, since currently the `shuffle` option will be turned off if `rect` is used. This is problematic for me because the images visualized in Weights and Biases will be from the same collection/with the most extreme aspect ratio. What would not sorting by aspect ratio affect? Thanks!

**Comment by glenn-jocher** [comment]
@yangsiyu007 I don't understand your question. shuffle is enabled for train, aspect ratio sorted in val for speed.

**Comment by yangsiyu007** [comment]
@glenn-jocher I'm wondering why sorting by aspect ratio would speed up inference? :)

**Comment by glenn-jocher** [comment]
@yangsiyu007 you don't have to ask me, you can set rect=False in the val.py dataloader and profile both ways
